server/calls_targets.go

```diff
@@ -0,0 +1,87 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+	"time"
+
+	promModel "github.com/prometheus/common/model"
+
+	"github.com/mattermost/mattermost/server/public/model"
+)
+
+func resolveURL(u string, timeout time.Duration) ([]net.IP, string, error) {
+	parsed, err := url.Parse(u)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to parse url: %w", err)
+	}
+
+	host, port, err := net.SplitHostPort(parsed.Host)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to split host/port: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	ips, err := net.DefaultResolver.LookupIP(ctx, "ip4", host)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to lookup ips: %w", err)
+	}
+
+	return ips, port, nil
+}
+
+func (p *Plugin) generateCallsTargets(appCfg *model.Config, host, port string, nodes []*model.ClusterDiscovery) ([]promModel.LabelSet, error) {
+	// First, figure out if Calls is running. If so, add the plugin metrics endpoint to the targets.
+	// Also, check if the external RTCD service is configured, in which case add its endpoints to targets.
+	status, err := p.API.GetPluginStatus(callsPluginID)
+	if err != nil {
+		return nil, fmt.Errorf("generateCallsTargets: failed to get calls plugin status: %w", err)
+	}
```

Comment:
One issue I foresee is that in the single node case we only generate targets once at the activation. This can cause problems if Calls starts after this plugin during the startup sequence. It's racy anyway. We may need to either do some polling or have Calls side signal it's ready somehow (needs changes on the other side though, so not ideal).

Reply:
Configuring it only once isn't a problem, scrape manager will continuously try to scrape the target if that's the concern. Or is it something else?

Reply:
@isacikgoz The problem I see is the generation of targets which I think is only done once (when not HA). If the Calls plugin is not fully running by then (it can happen during server initialization) then we could end up missing it completely.

Reply:
Oh gotcha, that's a likely scenario indeed. Then it would make a lot of sense to discover (eg. polling) plugins continuously.